python/cudf_polars/cudf_polars/experimental/scheduler.py

```diff
@@ -0,0 +1,171 @@
+# SPDX-FileCopyrightText: Copyright (c) 2025, NVIDIA CORPORATION & AFFILIATES.
+# SPDX-License-Identifier: Apache-2.0
+"""Synchronous task scheduler."""
+
+from __future__ import annotations
+
+from collections import defaultdict
+from collections.abc import MutableMapping
+from typing import TYPE_CHECKING, Any, TypeVar
+
+from typing_extensions import Unpack
+
+if TYPE_CHECKING:
+    from collections.abc import Mapping
+    from typing import TypeAlias
+
+
+Key: TypeAlias = str | tuple[str, Unpack[tuple[int, ...]]]
+Graph: TypeAlias = MutableMapping[Key, Any]
+T_ = TypeVar("T_")
+
+
+# NOTE: This is a slimmed-down version of the single-threaded
+# (synchronous) scheduler in `dask.core`.
+#
+# Key Differences:
+# * We do not allow a task to contain a list of key names.
+#   Keys must be distinct elements of the task.
+# * We do not support nested tasks.
+
+
+def istask(x: Any) -> bool:
+    """Check if x is a callable task."""
+    return isinstance(x, tuple) and bool(x) and callable(x[0])
+
+
+def is_hashable(x: Any) -> bool:
+    """Check if x is hashable."""
+    try:
+        hash(x)
+    except TypeError:
+        return False
+    else:
+        return True
+
+
+def _execute_task(arg: Any, cache: Mapping) -> Any:
+    """Execute a compute task."""
+    if istask(arg):
+        return arg[0](*(_execute_task(a, cache) for a in arg[1:]))
+    elif is_hashable(arg):
+        return cache.get(arg, arg)
+    else:
+        return arg
+
+
+def required_keys(key: Key, graph: Graph) -> list[Key]:
+    """
+    Return the dependencies to extract a key from the graph.
+
+    Parameters
+    ----------
+    key
+        Root key we want to extract.
+    graph
+        The full task graph.
+
+    Returns
+    -------
+    List of other keys needed to extract ``key``.
+    """
+    maybe_task = graph[key]
+    return [
+        k
+        for k in (
+            maybe_task[1:]
+            if istask(maybe_task)
+            else [maybe_task]  # maybe_task might be a key
+        )
+        if is_hashable(k) and k in graph
+    ]
```

Comment:
For large task graphs this could benefit from being written as a set intersection. This function does not need to be stable w.r.t. input order, right? Dependencies of a task can be executed in any order. The result should still be a valid topological sort, but not necessarily in the canonical ordering. If that makes debugging etc harder then we can leave it as is or consider rolling a small implementation of an ordered set (which is fairly trivial to do by just wrapping a dict).

Reply:
I think this is probably not worth it. A nicer approach (which we may move to) is to define tasks and dependencies as actual types, rather than relying on introspection like this.

Reply:
>I think this is probably not worth it. A nicer approach (which we may move to) is to define tasks and dependencies as actual types, rather than relying on introspection like this.

I'd definitely like to do this in a follow-up. The purpose of this PR is to essentially "copy" the single-threaded Dask scheduler into `cudf_polars` in the interest of avoiding a Dask dependency for single-GPU execution. However, I think we definitely want our own Task/Graph specification that can be (optionally) mapped onto the Dask specification for multi-gpu execution. In the future, we may be able to use rapidsmpf for multi-gpu scheduling.

Reply:
SGTM.